### ISSUE TITLE
Responsive Design, floating pixel size

### ISSUE THREAD

**Issue by Kyezil** [gfxfundamentals/threejsfundamentals#28]
Thanks for this great lesson.
The code is currently using `clientWidth` and `clientHeight` to get canvas display size.
In my browser I had a canvas with floating point size and because the properties above return long integers I had a scrollbar appear with a very small margin. Concretely my canvas size was 864.5 and the function would set it to 865.

I solved the problem by using `getBoundingClientRect()`:
```
const canvas = renderer.domElement;
const area = canvas.getBoundingClientRect()
const pixelRatio = window.devicePixelRatio;
const width  = area.width  * pixelRatio | 0;
const height = area.height * pixelRatio | 0;
```

**Comment by greggman** [comment]
I'm not saying you didn't run into an issue but  864.5 | 0 = 864

<img width="130" alt="Screen Shot 2019-07-10 at 10 01 29" src="https://user-images.githubusercontent.com/234804/60932596-b6445a00-a2f9-11e9-8737-ff0717e06a67.png">

Are you sure the issue wasn't something else?   Or maybe I don't understand what you mean

**Comment by greggman** [comment]
Actually thinking about this more scrollbars showing up suggests your CSS is set wrong.

The code in the article doesn't care what size the canvas is. You can set the canvas to `canvas.width = 4000` and you still shouldn't get scrollbars. 

Are you are your CSS is set correctly?

**Comment by Kyezil** [comment]
Thanks for looking into this. I haven't properly explained myself.
I was passing `true` to `renderer.setSize` to double check that the canvas had correct size.
I looked more into it and when my height is 864.5 (in computed tab of inspector) `clientHeight` is 865.
Changing it to `getBoundingClientRect` would give me a height of 864 (as you mention) instead of 865.

 I just noticed it when scrollbars appeared using `clientHeight` due to the rounding up and not when using `getBoundingClientRect` as `| 0` would round down in that case.

So it's a rounding issue, probably browser dependent. And with `false` to `renderer.setSize`, it's at most 1 pixel size difference in the canvas size depending on the option. I am not sure how relevant that is.
